File: pkmswitch100/pkmswitch100.rb
#!/usr/bin/env ruby

def optParser
  opts = {}
  ARGV.each do |arg|
    arg_arr = arg.split('=')
    opts[arg_arr[0].delete('-')] = arg_arr[1]
  end
  opts
end

def vercomp(old = '', new = '')
  # 3.2.2-3.4 3.2.2-3.3 3.2.2-1.1
  ov = old.gsub(/-.*$/, '')
  nv = new.gsub(/-.*$/, '')
  orel = old.gsub(/^.*-/, '').to_f
  nrel = new.gsub(/^.*-/, '').to_f
  # compare major version
  if nv.to_f > ov.to_f
    return 1
  elsif nv.to_f < ov.to_f
    return -1
  else
    # if the first 2 digits are the same, compare the last digit to make sure
    ovn = ov.gsub(/^.*\./, '').to_i
    nvn = nv.gsub(/^.*\./, '').to_i
    if nvn > ovn
      return 1
    elsif nvn < ovn
      return -1
    else
      # need to compare release numbers
      if nrel > orel
        return 1
      elsif nrel < orel
        return -1
      else
        return 0
      end
    end
  end
end

def max(arr = [])
  size = arr.length
  max = arr[size - 1]
  size.times do |i|
    max = arr[i] if vercomp(arr[i], max) < 0
  end
  max
end

def min(arr = [])
  size = arr.length
  min = arr[size - 1]
  size.times do |i|
    min = arr[i] if arr[i] < min
  end
  min
end

ffmpeg3_arr = %w(ffmpeg libavcodec57 libavdevice57 libavfilter6 libavformat57 libavresample3 libavutil55 libpostproc54 libswresample2 libswscale4)
vlc_arr = ['libvlc5', 'libvlccore8', 'vlc', 'vlc-codec-gstreamer', 'vlc-noX', 'vlc-qt', 'vlc-codecs', 'npapi-vlc']
gstreamer_arr = ['gstreamer-plugins-bad', 'gstreamer-plugins-libav', 'gstreamer-plugins-ugly', 'gstreamer-plugins-ugly-orig-addon', 'libgstadaptivedemux-1_0-0', 'libgstbadaudio-1_0-0', 'libgstbadbase-1_0-0', 'libgstbadvideo-1_0-0', 'libgstbasecamerabinsrc-1_0-0', 'libgstcodecparsers-1_0-0', 'libgstgl-1_0-0', 'libgstmpegts-1_0-0', 'libgstphotography-1_0-0', 'libgsturidownloader-1_0-0', 'libgstwayland-1_0-0', 'gstreamer-plugins-bad-orig-addon', 'libgstinsertbin-1_0-0', 'libgstplayer-1_0-0', 'libgstvdpau']
all_arr = ffmpeg3_arr + vlc_arr + gstreamer_arr

opts = optParser
pkgs = {}
all_arr = ffmpeg3_arr if opts.key?('ffmpeg')
all_arr = vlc_arr if opts.key?('vlc')
all_arr = gstreamer_arr if opts.key?('gstreamer')

pool_size = all_arr.length + 1
jobs = Queue.new
all_arr.each { |i| jobs.push i }

workers = Array.new(pool_size) do
  Thread.new do
    begin
      while x = jobs.pop(true)
        stdin = `LANG=en_US.UTF-8 zypper --no-refresh se -v #{x}`
        stdin.each_line do |line|
          next unless line =~ /^(i(\+)?|v)\s+\|/ && line.index(x + "\s")
          # v | libavformat57   | package | 3.2.2-3.4 | i586   | packman
          arr = line.split('|').each(&:strip!)
          arr[0] = "i" if arr[0] == 'i+'
          if pkgs.key?(arr[1])
            if pkgs[arr[1]].key?(arr[0])
              if pkgs[arr[1]][arr[0]].key?(arr[3])
                pkgs[arr[1]][arr[0]][arr[3]] << arr[4]
              else
                pkgs[arr[1]][arr[0]][arr[3]] = [arr[4]]
              end
            else
              pkgs[arr[1]][arr[0]] = { arr[3] => [arr[4]] }
            end
          else
            pkgs[arr[1]] = { arr[0] => { arr[3] => [arr[4]] } }
          end
        end
      end
    rescue ThreadError
    end
  end
end

workers.map(&:join)

oss = []
update = []
pkgs.each do |k, v|
  instdver = v['i'].keys[0]
  pkmver = max(v['v'].keys).to_s
  if v['v'].keys.include?(instdver)
    # not from packman
    oss << k if instdver != pkmver
  else
    # if the installed package's release number is near to packman's
    # then you have outdated packman version. if near to oss's
    # then it is oss package.
    instdrel = instdver.gsub(/^.*-/, '')
    vers = v['v'].keys.map { |i| i.gsub(/^.*-/, '') }
    diff = vers.map { |j| j.to_f - instdrel.to_f }.map do |m|
      m < 0 ? m * -1 : m
    end
    mini = min(diff)
    pos = diff.find_index(mini)
    near = v['v'].keys[pos]
    if near == pkmver
      update << k
    else
      oss << k
    end
  end
end

puts '======================= Packages not from Packman ========================='

if oss.empty?
  puts 'Good! All packages are from Packman!'
else
  str = ''
  oss.each do |i|
    puts i
    # zypper install libavformat57-3.2.2-3.4.x86_64
    pkm = max(pkgs[i]['v'].keys).to_s
    arch = pkgs[i]['i'].values[0][0]
    full = i + '-' + pkm + '.' + arch
    str = str + ' ' + full
  end
  puts "\nFIX: Run 'sudo zypper install" + str + "'."
end

puts '======================= Packman Packages need updates ====================='

if update.empty?
  puts 'Good! All packages are from Packman and at their latest versions!'
else
  str = ''
  update.each do |i|
    puts i
    str = str + ' ' + i
  end
  puts "\nFIX: Run 'sudo zypper up" + str + "'."
end

File: pkmswitch100/pkmswitch100.go
package main

import (
	"flag"
	"fmt"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"

	"github.com/PuerkitoBio/goquery"
	"github.com/gookit/color"
	"github.com/marguerite/diagnose/zypp/repository"
	"github.com/marguerite/diagnose/zypp/search"
	"github.com/marguerite/go-stdlib/httputils"
	system "github.com/marguerite/go-stdlib/runtime"
	"github.com/marguerite/go-stdlib/slice"
	semver "github.com/openSUSE-zh/node-semver"
)

var (
	ffmpeg    = map[string]int{"libavcodec": 57, "libavdevice": 57, "libavfilter": 6, "libavformat": 57, "libavresample": 3, "libavutil": 55, "libpostproc": 54, "libswresample": 2, "libswscale": 4}
	vlc       = []string{"libvlc5", "libvlccore9", "vlc", "vlc-codec-gstreamer", "vlc-noX", "vlc-qt", "vlc-codecs", "vlc-vdpau"}
	gstreamer = []string{"gstreamer-plugins-bad", "gstreamer-plugins-bad-chromaprint", "gstreamer-plugins-bad-fluidsynth", "gstreamer-plugins-bad-orig-addon", "gstreamer-plugins-libav", "gstreamer-plugins-ugly", "gstreamer-plugins-ugly-orig-addon", "gstreamer-transcoder", "libgstadaptivedemux-1_0-0", "libgstbadaudio-1_0-0", "libgstbasecamerabinsrc-1_0-0", "libgstcodecparsers-1_0-0", "libgstcodecs-1_0-0", "libgstinsertbin-1_0-0", "libgstisoff-1_0-0", "libgstmpegts-1_0-0", "libgstphotography-1_0-0", "libgstplayer-1_0-0", "libgstsctp-1_0-0", "libgsttranscoder-1_0-0", "libgsturidownloader-1_0-0", "libgstvulkan-1_0-0", "libgstwayland-1_0-0", "libgstwebrtc-1_0-0"}
)

func main() {
	var typ string
	flag.StringVar(&typ, "type", "all", "which package set to check, available: ffmpeg, vlc, gstreamer, all.")
	flag.Parse()

	// automatically check ffmpeg updates
	avcodec := latestPackmanVersion("libavcodec", 2)
	idx := avcodec - ffmpeg["libavcodec"]
	if idx > 0 {
		for k, v := range ffmpeg {
			ffmpeg[k] = v + idx
		}
	}

	// automatically check vlccore updates
	vlccore := latestPackmanVersion("libvlccore", 1)
	if vlccore > 9 {
		for i, v := range vlc {
			if v == "libvlccore9" {
				tmp := append(vlc[:i], "libvlccore"+strconv.Itoa(vlccore))
				vlc = append(tmp, vlc[i+1:]...)
			}
		}
	}

	// prepare the candidates
	var all []string
	for k, v := range ffmpeg {
		all = append(all, k+strconv.Itoa(v))
	}
	if typ != "all" {
		switch typ {
		case "vlc":
			all = vlc
		case "gstreamer":
			all = gstreamer
		}
	} else {
		slice.Concat(&all, vlc)
		slice.Concat(&all, gstreamer)
	}

	// initialize arch and repository
	arch := "i586"
	if strings.HasSuffix(runtime.GOARCH, "64") {
		arch = "x86_64"
	}

	repo := "packman"
	repositories := repository.NewRepositories()
	for _, v := range repositories {
		if strings.Contains(v.BaseURL, "packman") {
			repo = v.Name
		}
	}

	var notInstalled []string
	var notFromPackman []string
	needUpdate := make(map[string]struct{})

	wg := sync.WaitGroup{}
	wg.Add(len(all))
	mux := sync.Mutex{}

	for _, v := range all {
		go func(str string) {
			defer wg.Done()
			searchables := search.NewSearch(str, false).FilterByArch(arch)
			if len(searchables) == 0 {
				mux.Lock()
				notInstalled = append(notInstalled, str)
				mux.Unlock()
				return
			}

			installed := searchables.Installed()

			if len(installed.FilterByRepository(repo)) == 0 && installed[0].Repository != "(System Packages)" {
				mux.Lock()
				notFromPackman = append(notFromPackman, str)
				mux.Unlock()
				return
			}

			sv := semver.NewSemver(installed[0].Version)

			for _, v1 := range searchables.Available() {
				sv1 := semver.NewSemver(v1.Version)
				if sv1.GreaterThan(sv) {
					// only check packman updates not oss or others
					if v1.Repository == repo {
						if _, ok := needUpdate[str]; !ok {
							needUpdate[str] = struct{}{}
						}
					}
				}
			}

		}(v)
	}

	wg.Wait()

	allDone := true

	if len(notInstalled) > 0 {
		color.Info.Println("====== Packages not installed ======")
		color.Error.Println(strings.Join(notInstalled, " "))
		fmt.Println("FIX: sudo zypper in " + strings.Join(notInstalled, " ") + " --from " + repo)
		allDone = false
	}
	if len(notFromPackman) > 0 {
		color.Info.Println("====== Packages not from Packman ======")
		color.Error.Println(strings.Join(notFromPackman, " "))
		fmt.Println("FIX: sudo zypper in " + strings.Join(notFromPackman, " ") + " --from " + repo)
		allDone = false
	}
	if len(needUpdate) > 0 {
		color.Info.Println("====== Pakcages should be updated ASAP ======")
		var strs []string
		for k := range needUpdate {
			strs = append(strs, k)
		}
		color.Error.Println(strings.Join(strs, " "))
		fmt.Println("FIX: sudo zypper up " + strings.Join(strs, " "))
		allDone = false
	}

	if allDone {
		color.Info.Println("Good! All packages are from Packman and at their latest versions!")
	}
}

func latestPackmanVersion(pkg string, num int) int {
	suseVersion := system.LinuxDistribution()
	uri := "http://mirrors.hust.edu.cn/packman/suse/" + strings.ReplaceAll(suseVersion, " ", "_") + "/Essentials/x86_64/"

	c := httputils.ProxyClient()
	resp, err := c.Get(uri)
	defer resp.Body.Close()
	if err != nil {
		panic(err)
	}
	if resp.StatusCode != 200 {
		panic("response code not 200")
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		panic(err)
	}

	var hrefs []int

	doc.Find("pre a").Each(func(i int, s *goquery.Selection) {
		href, _ := s.Attr("href")
		if strings.HasPrefix(href, pkg) {

			j, err := strconv.Atoi(strings.TrimPrefix(href, pkg)[:num])
			if err != nil {
				panic(err)
			}
			hrefs = append(hrefs, j)
		}
	})
	sort.Ints(hrefs)
	return hrefs[len(hrefs)-1]
}
